_posts/2020-12-18-sensesp-1.md

```diff
@@ -0,0 +1,46 @@
+
+![IMG_20201217_234130](https://user-images.githubusercontent.com/1049678/102636535-7378e980-415d-11eb-9768-0c90d1856176.jpg)
+
+_A custom-built ESP32 development board in a waterproof enclosure, all ready waiting for new aquatic adventures!_ 
+
+The previous release, 0.5.1, was done on Nov 4. Since then, some of the notable new features include:
+
+- Support for Signal K metadata. SensESP can now provide information such as units, names, and description of the paths it is producing values for.
+- Support for both incoming and outgoing [Signal K PUT requests](https://signalk.org/specification/1.5.0/doc/put.html) (can both control other devices and receive commands)
+- A new LambdaConsumer for creating low-overhead consumer classes
+System status producer for implementing better status displays
+- Support for button UIs with debouncing and double/long click detection
+- Support for smart switches (networked remote controlled switches)
+- Support for RGB leds
+- Improvements in how the software is configured and initialized
+- Awesome new interactive Doxygen [class documentation](https://signalk.org/SensESP/generated/docs/annotated.html)
+
+Additionally, a LOT of bugs have been fixed.
+
+![grafana_engine](https://user-images.githubusercontent.com/1049678/102636567-7d025180-415d-11eb-9cfd-227ff4ba837c.png)
+
+_SensESP generated data in Grafana: engine RPM and oil, coolant, and exhaust temperatures_
+
+We definitely intend to continue the development. The complexity of the project has been a major pain point, with the build system breaking all too often. Also, code compilation takes a long time, and the software eats a lot of memory, both flash and RAM. These will be tackled by breaking up the monolith: hardware-specific components such as the BMx280 temperature sensors will be split away from the core software. In the future, they have to be explicitly included, but the gains in build times and build system stability will be well worth it.
+
+At the moment, SensESP is rather tightly integrated with Signal K: it only supports a Wifi data link layer with websockets and Signal K as the session and application layers. I and Joel have been musing about modularizing all this. It should be possible to use wired ethernet instead of Wifi; UDP broadcasts instead of Signal K; or even CAN bus and NMEA 2000 Signal K. Heck, it should be possible to have Signal K deltas over CAN bus! The basic building blocks of SensESP (the ReactESP reactive programming framework and the abstractions of observables, observers, producers, consumers, sensors, and transports) are easily flexible enough for that.
```

Comment:
The "or even CAN bus and NMEA 2000 Signal K" is my original text but doesn't actually make sense. I think the intent was "or even CAN bus and NMEA 2000.

I think the second CAN bus could actually be more specific it was NMEA 2000. My thinking was that there could be a SK-proprietary PGN that would encapsulate Signal K deltas. Then you could have, say, an engine temperature sensor that would output NMEA 2000 standard PGNs but also SK deltas over the same bus, providing more accurate information for SK-capable listeners.

Then again, if you want to avoid any implied NMEA 2000 compliance, I'm absolutely fine with that.

Reply:
See my latest commit. Changed UDP comparison to WebSocket to make more sense.